ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml

```diff
@@ -633,44 +642,57 @@ let time_limited_read filedesc length target_response_time =
   let total_bytes_to_read = length in
   let bytes_read = ref 0 in
   let buf = Bytes.make total_bytes_to_read '\000' in
-  let now = ref (Unix.gettimeofday ()) in
-  while !bytes_read < total_bytes_to_read && !now < target_response_time do
-    let remaining_time = target_response_time -. !now in
-    wait remaining_time ;
-    let bytes_to_read = total_bytes_to_read - !bytes_read in
-    let bytes =
-      try Unix.read filedesc buf !bytes_read bytes_to_read
-      with
-      | Unix.Unix_error (Unix.EAGAIN, _, _)
-      | Unix.Unix_error (Unix.EWOULDBLOCK, _, _)
-      ->
-        0
-    in
-    (* read into buffer=buf from offset=bytes_read, length=bytes_to_read *)
-    if bytes = 0 then
-      raise End_of_file (* End of file has been reached *)
-    else
-      bytes_read := bytes + !bytes_read ;
-    now := Unix.gettimeofday ()
+  while !bytes_read < total_bytes_to_read do
+    match Clock.Timer.remaining target_response_time with
+    | Expired _ ->
+        raise Timeout
+    | Remaining remaining_time ->
+        if not (wait remaining_time) then raise Timeout ;
+        let bytes_to_read = total_bytes_to_read - !bytes_read in
+        let bytes =
+          try Unix.read filedesc buf !bytes_read bytes_to_read
```

Comment:
Code seems almost all duplicated, only the function called changes (write / read), can this be factored out?

Reply:
Xapi-fdcaps has a better factored out implementation (there is a `repeat_read` and `repeat_write`, although much of the code of those 2 functions could be shared too, and there is no Timer version, but one could be easily introduced).

Xapi-fdcap is what I use to test unixext, so if we switch unixext over to xapi-fdcaps directly, then it would be testing itself with itself (although the way timeout testing is done is independent of how it is implemented, so the test would still be a useful one).

Unixext could be changed to use xapi-fdcaps, although the type changes will probably ripple through the codebase. It might result in an overall safer code, because we'd know from its type whether a file descriptor is a socket or not for example (and then whether calling setsockopt is a valid call or not).

I'll check whether switching over directly to fdcaps would result in simpler code, if not I'll factor out the code here in stdext.